### DISCUSSION TITLE
ACCESSOR_VECTOR3_NON_UNIT Error When Converting JSON Format gltf Files to Binary Format glb

### DISCUSSION THREAD

**Discussion by mnakaji1-avail** [donmccurdy/glTF-Transform#1467] [Q&A]
I make a program that loads a set of JSON glTF files, manipulates them using glTF Transform, and downloads the binary glb files. It reads the each files using FileReader, creates a JSONDocument, makes it a Document using WebIO.readJSON(), and then unpartitions() .
`const jsonDocument: JSONDocument = {
  json: {} as GLTF.IGLTF,
  resources: {},
};
//... load gltf files...
const webIO = new WebIO({ credentials: "include" })
    .registerExtensions([...KHRONOS_EXTENSIONS]);
await document.transform(unpartition());
//..document to glb and download
`
i use gltf.report to check the resulting glb file, i got many ACCESSOR_VECTOR3_NON_UNIT errors. On the other hand, when i download the document as a glTF file using partition() instead of glb, i didn't get a validation error. How can I make the gltf file glb so that it does not get a validation error?

validation errors sample
ID | CODE | MESSAGE | SEVERITY | POINTER
-- | :-- | :-- | :-- | :--
0 | ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 4640..4642 is not of unit length: 0. | 0 | /meshes/2/primitives/0/attributes/TANGENT
1 | ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 4644..4646 is not of unit length: 0. | 0 | /meshes/2/primitives/0/attributes/TANGENT
2 | ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 4648..4650 is not of unit length: 0. | 0 | /meshes/2/primitives/0/attributes/TANGENT
3 | ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 4652..4654 is not of unit length: 0. | 0 | /meshes/2/primitives/0/attributes/TANGENT
4 | ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 4656..4658 is not of unit length: 0. | 0 | /meshes/2/primitives/0/attri

errors continues.

Thanks for any tips.

**Comment by donmccurdy** [comment]
I suspect the contents of the .gltf and .glb files are identical, but that perhaps the validation is not checking the binary data in the external .bin of the .gltf. Could look more closely if files could be attached as a .ZIP, but I suspect this is not the issue.

Because the model contains zero-length vertex tangents, two options would be:

- A. remove the tangents
    ```javascript
    for (const mesh of document.getRoot().listMeshes()) {
      for (const prim of mesh.listPrimitives()) {
        const tangent = prim.getAttribute('TANGENT');
        if (tangent) tangent.dispose();
      }
    }

    ```
- B. rebuild the vertex attribute with [`tangents()`](https://gltf-transform.dev/modules/functions/functions/tangents)

If you're not sure that you need the tangents, I would go with (a). Most models do not need them in my experience, but they're helpful as a fix if you see certain problems with a normal map.

**Reply by mnakaji1-avail** [reply]
donmccurdy Thank you very much. I understood with your kind explanation. That means that glb and gltf have the same contents, but bin in gltf just was not validated. I didn't get the error of ACCESSOR after I disposed the tangent in the way of A.

It seems that there are still other warnings, but this is a different problem, so I will check it by myself first. IMAGE_FEATURES_UNSUPPORTED

I am very happy to receive a direct answer from you, I know you are a great person in 3D related technology. Thank you very much. It was very helpful.
